components/remote_settings/src/client.rs

```diff
@@ -102,22 +101,27 @@ impl Client {
     /// Downloads an attachment from [attachment_location]. NOTE: there are no
     /// guarantees about a maximum size, so use care when fetching potentially
     /// large attachments.
-    pub fn get_attachment(&self, attachment_location: &str) -> Result<Response> {
-        let mut current_remote_state = self.remote_state.lock();
-        self.ensure_no_backoff(&mut current_remote_state.backoff)?;
-        let attachments_base_url = match &current_remote_state.attachments_base_url {
-            Some(url) => url.clone(),
-            None => {
-                let req = Request::get(self.base_url.clone());
-                let server_info = req.send()?.json::<ServerInfo>()?;
-                match server_info.capabilities.attachments {
-                    Some(capability) => Url::parse(&capability.base_url)?,
-                    None => Err(RemoteSettingsError::AttachmentsUnsupportedError)?,
-                }
-            }
+    pub fn get_attachment(&self, attachment_location: &str) -> Result<Vec<u8>> {
+        Ok(self.get_attachment_raw(attachment_location)?.body)
+    }
+
+    /// Fetches a raw network [Response] for an attachment.
+    pub fn get_attachment_raw(&self, attachment_location: &str) -> Result<Response> {
+        if let Some(attachments_base_url) = self.remote_state.lock().attachments_base_url.clone() {
+            return self.make_request(attachments_base_url.join(attachment_location)?);
+        }
+
+        let server_info = self
+            .make_request(self.base_url.clone())?
+            .json::<ServerInfo>()?;
+        let attachments_base_url = match server_info.capabilities.attachments {
+            Some(capability) => Url::parse(&capability.base_url)?,
+            None => Err(RemoteSettingsError::AttachmentsUnsupportedError)?,
         };
-        current_remote_state.attachments_base_url = Some(attachments_base_url.clone());
-        drop(current_remote_state);
+        self.remote_state
+            .lock()
+            .attachments_base_url
+            .get_or_insert(attachments_base_url.clone());
```

Comment:
This also confused me and the semantics are subtly changed. On one hand this is better because it doesn't hold the lock while we make a request, but OTOH, now it seems possible multiple threads will race to update the base URL and the first will win. If the expectation is that the server will never update the URL (which are the semantics pre this patch IIUC), should we assert/log/err the existing value is the same once we've re-acquired the lock? Or if we want to handle the face that the server might update the base URL, should we use `Option::insert()` to ensure the last request wins?

(TBH this is such an edge-case I don't particularly care, but maybe some comments with the implicit assumptions might help)